File: web/src/game/combo.ts
import { generateHint } from './hints'
import type { ComboStage, Difficulty } from './types'

export const DIAL_SIZE = 100
export const STAGE_COUNT = 3
export const ROUND_SECONDS = 180

// How long the dial must sit still before a digit auto-confirms. This
// encoder has no physical confirm button, so "confirm" is a dwell timeout.
export const DWELL_MS = 1000

// Time docked from the clock for a wrong dwell-confirm, to make
// guess-and-check spinning through every number a losing strategy.
export const PENALTY_SECONDS = 15

const randInt = (min: number, max: number): number =>
  Math.floor(Math.random() * (max - min + 1)) + min

export function generateRound(difficulty: Difficulty): ComboStage[] {
  return Array.from({ length: STAGE_COUNT }, () => {
    const target = randInt(0, DIAL_SIZE - 1)
    return { hint: generateHint(target, difficulty), solved: false, wrongAttempts: 0 }
  })
}

File: standalone/game/combo.py
# Ported from web/src/game/combo.ts. No DWELL_MS here — this build has a
# real Confirm button, so there's no dwell-timeout to configure.

import random

from game.hints import generate_hint
from game.types import ComboStage

DIAL_SIZE = 100
STAGE_COUNT = 3
ROUND_SECONDS = 180

# Time docked from the clock for a wrong confirm, to make guess-and-check
# spinning through every number a losing strategy.
PENALTY_SECONDS = 15


def generate_round(difficulty):
    stages = []
    for _ in range(STAGE_COUNT):
        target = random.randint(0, DIAL_SIZE - 1)
        stages.append(ComboStage(generate_hint(target, difficulty)))
    return stages
